File: Data Structures & Algorithms/products-of-array-discluding-self/submission-11.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len=nums.length;
        int[] lToRProduct=new int[len];
        int[] rToLProduct=new int[len];
        int[] res=new int[len];
        lToRProduct[0]=1;
        rToLProduct[len-1]=1;
        for(int i=1, j=len-2; i<len; i++, j--){
            lToRProduct[i]=lToRProduct[i-1]*nums[i-1];
            rToLProduct[j]=rToLProduct[j+1]*nums[j+1];
        }
        for(int i=0; i<len; i++){
            res[i]=lToRProduct[i]*rToLProduct[i];
        }
        return res;
    }
}  

File: Data Structures & Algorithms/top-k-elements-in-list/submission-11.rs
impl Solution {

    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let max_freq=nums.len();
        let mut freqs:HashMap::<i32, usize>=HashMap::new();
        let mut buckets:Vec<Vec<i32>>=vec![Vec::<i32>::new();max_freq+1];
        let mut res:Vec::<i32>=Vec::new();
        for num in nums{
            *freqs.entry(num).or_insert(0)+=1;
        }
        for (k, freq) in freqs{
            buckets[freq].push(k);
        }
        for i in (0..max_freq+1).rev(){
            for &k in buckets[i].iter(){
                res.push(k);
            }
            if res.len()==(k as usize){
                return res;
            }
        }
        return res;
    }
}
